### ISSUE TITLE
Get rid of knitr?

### ISSUE THREAD

**Issue by jarioksa** [vegandevs/vegan#279]
`vignettes/FAQ-vegan.Rmd` is written in R markdown and for this purpose we suggest **knitr** and declare **knitr** as a VignetteBuilder. the **knitr** package has many dependences, and therefore we end up installing a large number of other packages to build that one vignette: If we only look at imports (and ignore suggests), we end up installing **evaluate**, **glue**, **highr**, **magrittr**, **markdown**, **mime**, **stringi**, **stringr** and **yaml**. 

Is it worthwhile to manually translate Rmarkdown to Sweave to get rid of this chain of dependency?

**Comment by gavinsimpson** [comment]
I personally don't see the benefit of switching to a verbose markup system just to cut out some very popular and hence widely installed/used packages.

**Comment by jarioksa** [comment]
@gavinsimpson What would you think if you try to check a package in, say, **R** 3.2.0, and `R CMD check` fails because a `VignetteBuilder` is not installed, and within **R** `install.packages()` says that **knitr** is not available for this version of **R**?

**Comment by gavinsimpson** [comment]
I'd say "upgrade your R" :-) How old is 3.2.0 now? Must be at least 2 years

**Comment by psolymos** [comment]
I can relate to both sides (upgrade vs. keep it lean), but #282 point out that practicality matters more than principles.

This is the only Rmd file among the vignettes, and it has no evaluated code but only formatting. Rendering R markdown to the intermediate tex seems like an easy option. (I did it on my machine and there were 200 lines of markup definitions in the preamble which will have to go.)

**Comment by jarioksa** [comment]
@gavinsimpson : I don't know where **we** say that we should work with old **R**, but **I**'ve said it many times, and I think that we should be able to work at least with the old-release and one version before that. This means that 3.3.x should be OK. In no case we should have a blocking feature bound to current release, but previous should be OK. I have delayed implementing some features till they are old enough in **R**. However, I don't want to jump over two major versions, but this can be done when justified. In this `sigma.cca` case I just did not think that `sigma` is such an important user-level function that alone would justify moving up to 3.3.0 when there is an easy workaround which does not influence 3.3.0+ users.

As to ordination code, it was written so that support functions mostly work like earlier. Even direct access to ordination objects works  mostly like earlier. For the cases where the result object changed and direct access failed, we provided compatibility functions already in **vegan** 2.4.x and alerted the package authors to switch to these functions even before we changed to the new structure. See `ordiYbar` code. I still provide the compatibility layer after moving to 2.5.x, and I also made pull requests to some of these packages to fix them so that they will work both with the old, current and future **vegan**. Without these fixes, they would only break in **vegan** 2.6.x, but not yet (and now they are fixed).

**Comment by jarioksa** [comment]
The main issue here was **knitr**.  I still don't like its dependency net, but I stop here (I wrote a longer rant, but deleted my jeremiad). 

I was only frustrated when I was not able to install **knitr** with its dependencies automatically using `install.packages()`, and setting **R** environmental variables or switches like `--no-build-vignettes`, `--no-vignettes` did not help, but checking failed because **knitr** was still needed. It looked like the only alternative would be to download **knitr** source + sources of all its dependencies and installing them manually in correct order. (I found a dirty trick to fool **R** believe I had installed **knitr** although I hadn't and then check worked with `--no-*vignettes` switches).

To my surprise the failure to install **knitr** was not due to to `install.packages`(that claimed that **knitr** source file is not available for this version of **R**), but it was due to RStudio cloud. I have lately preferred to use cloud service to install packages, because it just feels a sensible way to do. However, for some reason that service checks and rejects **R** versions. If I instead select a specific physical server, everything is OK, and **knitr** and (most importantly) packages it depends on are installed automatically.

In conclusion, I'll keep the Rmarkdown vignette (I hope they find out a leaner way of handling that file, though).

**Comment by gavinsimpson** [comment]
That sounds like a sync issue - I've never run into an issue building knitr via `install.packages()` and as it is used with pretty much the entire *tidyverse* I assume this was just a transitory hiccup.

(I've not bothered using my distro-provided R on Fedora/RHEL/Centos because it is never up to date and if I need to keep on top of various releases of R including R-devel it is just easier to build these things locally. Sticking with LTS or repo-provided packages for some open source software is not worth the hassle; I am in Pandoc purgatory for example because Fedora doesn't keep it's haskell stack up dated as fast as pandoc updates itself.)

Whilst this *knitr* things seems resolved (are you OK with the resolution of no changes now?) the broader issue is something that we perhaps should discuss or even you just state what the policy is so that those of us with write access here aren't doing things that do not align with your vision. This perhaps relates to how we might break `summary()` (so I can fix it and the `eigenvals()` issues for example) and when we want to do this?

**Comment by jarioksa** [comment]
@gavinsimpson : I have no problem with **knitr** if it can be installed easily in a freshly built **R**, and I think it can (and now I know how to trick **R** to believe it is installed even if it is not). I closed this issue because I think the situation is settled now.

However, `vignettes/FAQ-vegan.Rmd` contains no executed **R** code and we could do with plain markdown instead of Rmarkdown. I have no idea how to do that in a package and would anything change, though. I guess this would depend on pandoc and this would make it platform-fragile.

**Comment by jarioksa** [comment]
@gavinsimpson : It's not only I, but CRAN also requires that you shall not break other packages. Indeed, **vegan** 2.5-0 was rejected from CRAN because it broke **RAM** package. I had to fix this with 591318eb97e before **vegan** was accepted, and that was also the reason why CRAN never had 2.5-0 but the first **vegan** release was 2.5-1. 

Commit 591318eb97e actually patches `summary.cca` which highlights the importance of being careful with radical overhauls of functions used in other CRAN (or Bioconductor) packages. When you change `summary.cca`, you must run `tools::check_packages_in_dir()` to see if that breaks anything, and then work out how to get round the problem. I had run `tools::check_packages_in_dir()` for 2.5-0, but the installation failed for **RAM** (and several other packages) because they also depended on Bioconductor packages (usually **phyloseq**) that I did not have then (now I have), and I was caught by CRAN.

We have done quite a lot to avoid breaking other packages. For instance, after complete redesign of `nullmodel()` in 2.2-0, we still provided old `commsimulator()` which only wrapped `simulate(nullmodel())`. The `commsimulator` function was only deprecated in 2.5-0. The function was still used in two CRAN packages, but they didn't use it in their tests so that we passed CRAN. However, I made pull requests to these two packages (**bipartite** & **nodiv**) to fix the breakage we did.

Another example is the overhaul of constrained ordination which also brings along changes in the internal structure of the result object (issue #227). I found out that this breaks one CRAN package and is potentially dangerous to several scripts at large. So we proceeded in steps: first we made this a public issue (#227) and provided tools to handle both the old and new result objects already in 2.4-x (function `ordiYbar`: see its comments), and now in 2.5-x we have both the old and the new structures for compatibility. In 2.6-0 (current master branch) I removed the old structure. When I ran `tools::check_packages_in_dir()` for 2.6-0, I found out that the package we were about the break (**RVAideMemoire**) had changed its code to use `ordiYbar()` and was already OK, and the strategy was successful. We still break scripts, I'm sure, but if they follow **vegan** (for instance, read its release NEWS), they have had two or more years to adapt to the future.

**Comment by gavinsimpson** [comment]
@jarioksa That's not my understanding of of CRAN's policy. You can't just do what you want and make changes that break other packages causing CRAN people work without discussing this with them, without planning for the change, and without attempts (multiple) over a period of time to contact maintainers of dependencies that are broken due to changes.

The problem we had was not doing this process correctly (including me).

I'm going to revisit these things once I have my workflow set up to check reverse dependencies of *vegan*. I have it working for *permute*. This may also be something we can turn on periodically on Travis &mdash; I need to check but I think there is a way to do this, but it's not something we want to be running on every push.

**Comment by jarioksa** [comment]
@gavinsimpson : same message, different wording. I put it "you can't break CRAN packages". In this case I indeed failed because I could not test **RAM** and was unaware of it failing. I could not even claim that they had incorrectly assessed internal `summary.cca` structures -- they had, but using official accessor functions (`eigenvals.cca`) had the same problem. So I just got to fix **vegan** to get it to CRAN (and this is experimental knowledge).

That said, `summary.cca` should be easy to change smoothly, because it currently adds hardly nothing to `scores.cca` and `eigenvals.cca`. After breaking `summary.cca` we only need to find the packages that break, and ask their maintainers to use `scores` or `eigenvals` which will work before and after `summary.cca` change. Making a PR is a very efficient way to ask changes for those packages that are in github.

**Comment by jarioksa** [comment]
@gavinsimpson : I think it hardly makes sense to set up Travis testing of **vegan** dependencies. No doubt, it can be done, but it is hardly useful. Firstly, it is a huge operation taking a long time: installing and checking all packages depending on **vegan**  in CRAN and Bioconductor with their dependencies enhances climate change just too much. Most importantly, many packages will fail, but most not for **vegan**. What we really  are two tests with old and new versions **vegan** and then running `tools::check_packages_in_dir_changes()` to see if **vegan** caused *new* problems. Having all this in Travis uses still more energy and increases the global temperatures as much as flying from Finland to Canada.
